integration_test.go

```diff
@@ -0,0 +1,231 @@
+// (named by IntegrationServerFeatureToggle) is set by the calling test
+func TestIntegrationServer(t *testing.T) {
+	if os.Getenv(IntegrationServerFeatureToggle) != "1" {
+		t.Skip("Skipping as do not need to stand up integration test server in this run")
+	}
+
+	// Create new HTTP server
+	mux1 := mux.NewRouter()
+
+	mux1.HandleFunc(IntegrationServerPath, integrationTestServerHandler).
+		Methods("GET")
+
+	srv := NewServer("tcp", "localhost:"+IntegrationServerPort, mux1)
+	srv.Debug = true
+
+	// This is used to signal back to the calling test that the server has started up
+	fmt.Println("PID:", os.Getpid())
+	if err := srv.ListenAndServe(); err != nil {
+		fmt.Println(err)
+	}
+}
+
+// handler used by the integration test server
+func integrationTestServerHandler(w http.ResponseWriter, r *http.Request) {
+	duration, err := time.ParseDuration(r.FormValue("duration"))
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+	time.Sleep(duration)
+	fmt.Fprintf(
+		w, "Slept for %2v seconds. Response from pid %d.\n",
+		duration.Seconds(),
+		os.Getpid(),
+	)
+}
+
+// createIntegrationServer will re-run the test binary that is currently running,
+// but only execute the "TestIntegrationServer" test, which will stand up an
+// 'endless' server which can be used to test signal handling
+func createIntegrationServerTest() *exec.Cmd {
+	cmd := exec.Command(os.Args[0], "-test.run=TestIntegrationServer")
+
+	// Set the feature toggle for the integration test server
+	cmd.Env = []string{IntegrationServerFeatureToggle + "=1"}
+
+	return cmd
+}
+
+func TestIntegrationRestart(t *testing.T) {
+
+	intServer := createIntegrationServerTest()
+	serverMessages := make(chan string) // channel to receive notifications from the integration server
+	// Start test integration server
+	require.NoError(t, startIntegrationServer(intServer, serverMessages), "starting integration test server")
+	parentPid := intServer.Process.Pid
+	defer teardownServer(parentPid)
+	go intServer.Wait() // when this ends the parent process should be over and pipe clean up will happen
+
+	// The pid is printed out from the integation test server once it has fully started up
+	// Check for this message so we can start sending requests/signals to the server
+	msg, msgChannelOpen := <-serverMessages
+	require.True(t, msgChannelOpen, "Unexpected close of server messages channel")
+	msgPid, _ := strconv.Atoi(msg)
+	require.Equal(t, parentPid, msgPid, "Unexpected message down server messages channel")
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the server a chance to bind to port
+
+	// Send a long running request to the test server
+	longReqWait := &sync.WaitGroup{}
+	longReqWait.Add(1)
+	var (
+		longResp     *http.Response
+		longReqError error
+	)
+	go func() {
+		longResp, longReqError = sendIntegrationTestRequest(2 * time.Second)
+		longReqWait.Done()
+	}()
+
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the request a chance to get handled
```

Comment:
not needed as you have the waitgroup so use it?

Reply:
The wait group is for the long request to finish, I need to send the sighup before the the request finishes, and as the request is kicked off in a go routine I didn't want a race condition where the sighup arrived before the long request

Reply:
Agreed - I've tried to avoid them as much as possible, I guess I can add another message into the handler to recognise the request arriving, unless you can think of another solution?